### PR TITLE
[QOLDEV-297] Modify the url_for function to use the default dataset if a custom one fails

### PR DESCRIPTION
**Pull request by RossWebsterWork**
Copied from [[QOLSVC-492] Let trash links use the base /dataset type instead of custom types #72](https://github.com/qld-gov-au/ckan/pull/72)

[QOLSVC-492]: https://ssq-qol.atlassian.net/browse/QOLSVC-492?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Review by ThrawnCA — APPROVED** [review]

File: ckan/lib/helpers.py [hunk 1]
```diff
@@ -291,6 +291,70 @@ def _get_auto_flask_context():
...
+    Use ``qualified=True`` for a fully qualified URL when targeting a Pylons
+    endpoint.
+
+    For backwards compatibility, an effort is made to support the Pylons syntax
+    when building a Flask URL, but this support might be dropped in the future,
+    so calls should be updated.
+    '''
+    try:
+        return base_url_for(*args, **kw)
+    except FlaskRouteBuildError:
```
Comment (duttonw) [inline ckan/lib/helpers.py:332]: nice workaround.

**Comment by RossWebsterWork** [comment]
Tested on CKANTest_DEV and a deleted dataset (Test Delete) with a custom datatype successfully shows in Trash, but selecting the link throws a server error. I think this is trying to use the custom dataset type and doesn't use url_for(), which might need to be updated too to default to using the base dataset

**Review by ThrawnCA — APPROVED** [review]

**Comment by ThrawnCA** [comment]
> I think this is trying to use the custom dataset type and doesn't use url_for(), which might need to be updated too to default to using the base dataset

`ckan/templates/package/read_base.html` is using the custom `link_for` tag.

**Review by ThrawnCA — APPROVED** [review]

**Comment by RossWebsterWork** [comment]
I ran some tests locally for these changes by changing the package type in the local database, which threw internal server errors without the fix and displayed correctly with it, for both active and deleted datasets.

However, I did notice that datasets with the fix had a "/dataset" url instead of their dataset type, I am not sure if this is intended behaviour. Also, changing the dataset type to "geoscience" seemed to behave the same as setting it to "custom_dataset", both with the fix and without, so I am not sure that this method is an accurate way of testing different package types.

File: ckan/lib/helpers.py [hunk 2]
```diff
@@ -291,6 +290,71 @@ def _get_auto_flask_context():
...
+        # If the url failed, try again but replace any custom dataset type in
```
Comment (ThrawnCA) [inline ckan/lib/helpers.py:333]: Since there is a `return` inside the try block, we could just `pass` here and not have to indent everything.

**Review by ThrawnCA — APPROVED** [review]

File: ckan/lib/helpers.py (hunk 2 above)
Comment (RossWebsterWork) [reply ckan/lib/helpers.py:333]: I hadn't considered that and it almost works, but then the `raise` on line 353 wouldn't re-raise the FlaskRouteBuildError if it doesn't retry with default

**Pull request merged**